pkg/operator/controllers/cloudproviderconfig/cloudproviderconfig_controller.go

```diff
@@ -0,0 +1,239 @@
+package cloudproviderconfig
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers/base"
+)
+
+const (
+	ControllerName    = "CloudProviderConfig"
+	controllerEnabled = "aro.cloudproviderconfig.enabled"
+)
+
+var cloudProviderConfigName = types.NamespacedName{Name: "cloud-provider-config", Namespace: "openshift-config"}
+
+// / This is a very old version of the CloudProviderConfig struct from k8s upstream
```

Comment:
Given the history I'm not sure if there is much risk in this breaking in a few releases. 

This is more of a question of when OpenShift updates this API and we would want to pivot to use the new API. The other thing to note is that in the version of this struct we have in the OpenShift upstream, there are fields missing from that struct. Which would allow us to use the upstream k8s API. So it would be possible to bring in the upstream k8s API at that point. The missing fields are included in this struct I've provided here so it wont actually break.

To be clear this fork and divergence happened a long time ago... some time in 2018 from what I can gather. There is a large update to the cluster API going on in OpenShift right now so it is possible they update it soon but it might just stay the same too.

Reply:
Also if it isn't clear in the reply above the OpenShift cluster API actually moonlights at the upstream k8s API. The reason I can't just pull in the upstream k8s API is because the imports get confused since they both identify as the same Golang import module.

This works as expected when the OpenShift cluster API and the upstream k8s API are in sync but since the OpenShift cluster API is missing fields it causes errors when unmarshalling the JSON even if I pull in a newer version of k8s API that includes the fields.